### PR TITLE
[inductor] modify the output_stride of ConcatKernel

### PR DESCRIPTION
**Pull request by jiayisunx**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122761

Fix https://github.com/pytorch/pytorch/issues/121613.
Modify the `output_stride` of `ConcatKernel`: If any input to `Concat` is `Pointwise`, check the layout of all inputs to `Pointwise`, if any of the inputs is in channels_last format, set channels_last strides for the `output_stride`.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -3687,9 +3725,11 @@ def create(cls, inputs, dim):
 
         output_stride = FlexibleLayout.contiguous_strides(new_size)
         # If any of the inputs is in CL format, use CL format for the output
+        any_input_is_storage_and_layout = False
```
Comment (leslie-fang-intel) [inline torch/_inductor/ir.py:3728]: `any_input_is_storage_and_layout = any(is_storage_and_layout(x) for x in inputs)` looks clearer?

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -1817,6 +1817,44 @@ def as_storage_and_layout(x, freeze=True, want_contiguous=False, stride_order=No
     raise NotImplementedError
 
 
+def is_pointwise_with_channels_last_inputs(x):
+    if isinstance(x, TensorBox):
+        return is_pointwise_with_channels_last_inputs(x.data)
+    if isinstance(x, StorageBox) and isinstance(x.data, Pointwise):
+        all_reads = x.data.get_reads()
```
Comment (leslie-fang-intel) [inline torch/_inductor/ir.py:1824]: Similar as in https://github.com/pytorch/pytorch/pull/122760, we can get the buffer from `read.name`. Will it be easy to check the buffer's layout?

Comment (jgong5) [reply torch/_inductor/ir.py:1824]: Do I understand correctly that we can share some logic from #122760 too? Both need to get layouts of the input buffers.

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/ir.py [hunk 3]
```diff
@@ -3725,6 +3725,21 @@ def create(cls, inputs, dim):
                     # use CL stride for the output
                     output_stride = make_channels_last_strides_for(new_size)
                     break
+        any_input_is_storage_and_layout = any(is_storage_and_layout(x) for x in inputs)
+        fx_node_args = V.graph.current_node.args[0]
```
Comment (leslie-fang-intel) [inline torch/_inductor/ir.py:3729]: nit: Add an assertion to ensure that `V.graph.current_node` is a `aten.cat` node before accessing args[0].

File: torch/_inductor/ir.py [hunk 4]
```diff
@@ -3725,6 +3725,21 @@ def create(cls, inputs, dim):
...
+        if not isinstance(fx_node_args, (tuple, list)):
+            fx_node_args = [
+                fx_node_args,
```
Comment (leslie-fang-intel) [inline torch/_inductor/ir.py:3732]: Looks like we are handling the case `fx_node_args` is a single Tensor? Be curious in which scenario, the input of cat will be a single Tensor?

**Comment by leslie-fang-intel** [comment]
LGTM, just a few more comments.

File: torch/_inductor/ir.py [hunk 5]
```diff
@@ -3725,6 +3725,21 @@ def create(cls, inputs, dim):
...
+            ]
+        if any_input_is_storage_and_layout is False and any(
```
Comment (jgong5) [inline torch/_inductor/ir.py:3734]: suggest to add a note here explaining the heuristics we use here.

**Review by jgong5 — APPROVED** [review]

**Review by leslie-fang-intel — APPROVED** [review]

File: torch/_inductor/ir.py (hunk 3 above)
Comment (jiayisunx) [reply torch/_inductor/ir.py:3735]: Added, thanks!

File: torch/_inductor/ir.py (hunk 5 above)
Comment (jiayisunx) [reply torch/_inductor/ir.py:3739]: Added, thanks!

**Comment by jiayisunx** [comment]
@jansel , could you please review this PR? Thanks.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**